packages/adminrouter/extra/src/README.md

```diff
@@ -264,28 +264,178 @@ it, but it does not seem to be justified. For example:
+
+### Intended use
+It is very important to remember that `/service` endpoint was designed to work
+with root Marathon only. Tasks launched by MarathonOnMarathon instances will not
+be reachable via it.
+
+In the case of Mesos frameworks, this becomes a bit more complex - all the
+frameworks are visible in Mesos' `/state-summary` endpoint output so in theory
+it is possible to access a framework launched using MoM or by hand.
+Unfortunately, Mesos as of now does not enforce framework name uniqueness and
+thus it is possible to launch two frameworks with the same name. AR in such
+case will route in a deterministic manner. Root Marathon by default enforces
+unique names for all the tasks and frameworks it has under control.
+
+Please check the [Limitations](#limitations) section for more details.
+
+### Operation
+Under the hood, the `/service` endpoint follows non-trivial algorithm in order to
+resolve request path into IP + Port pair.
+
+#### Recursive resolving
+There is no way to tell which part of the path is namespace+service name and
+which one is just a resource address. For example, assumming that the request
+path is `/foo/bar/baz/picture.jpg` we can have:
+* namespace and name: `foo`, resource address: `/bar/baz/picture.jpg`
+* namespace and name: `foo/bar`, resource address: `/baz/picture.jpg`
+* namespace and name: `foo/bar/baz`, resource address: `/picture.jpg`
+* namespace and name: `foo/bar/baz/picture.jpg`, resource address: `/`
+
+So the algorithm iteratively tries to resolve each "candidate" namespace and name
+and assume that first-match wins.
+
+Assuming that we deployed `foo/bar` NGINX task, we will see two iterations:
+* first one with `foo` candidate namespace+service, which is not present in the
+cluster
+* second one with `foo/bar` candidate namespace+service name, which results in
+request being routed to correct backend. In case when user is trying to access
+an non-existent service, there are going to be four iterations, each one
+unsuccessful. After the fourth one the code will give up.
+
+There is of course a limit on the number of iterations that AR makes. Currently,
+it is set to `10`. This means that a service cannot nest more than nine times and
+that everything after the tenth segment is always treated as a resource component.
+
+#### Resolving individual component
+During each iteration described in previous paragraph, AR tries to resolve the
+component using the data stored in the cache and if necessary - from MesosDNS:
+* first step is to check if there is a root Marathon task that has lebels set
+  and these lebels:
+  * `DCOS_SERVICE_NAME` label contents is equal to given `namespace+service` string
+  * `DCOS_SERVICE_SCHEME` label is set
+  * `DCOS_SERVICE_PORT_INDEX` label is set
+  * task is in state `TASK_RUNNING`
+  If all the conditions above are met, request is routed using root Marathon
+  data. Please check the [section below](#recommended-way-to-expose-services)
```

Comment:
It is preferred (I prefer?) to use section names in links, rather than position. This means that someone moving the sections, or adding a section between these two, does not make the information here false.